Update documentation on sandboxes (#747)
huggingface/smolagents: docs/source/en/tutorials/secure_code_execution.mdx

This is why we put emphasis on proposing code agents, in this case python agents, which meant putting higher effort on building secure python interpreters.

### Local code execution??

By default, the `CodeAgent` runs LLM-generated code in your environment.

This is inherently risky, LLM-generated code could be harmful to your environment.
One could argue that on the [spectrum of agency](../conceptual_guides/intro_agents), code agents give much higher agency to the LLM on your system than other less agentic setups: this goes hand-in-hand with higher risk.

So you need to be mindful of security.

To add a first layer of security, code execution in `smolagents` is not performed by the vanilla Python interpreter.
We have re-built a more secure `LocalPythonInterpreter` from the ground up.

To be precise, this interpreter works by loading the Abstract Syntax Tree (AST) from your Code and executes it operation by operation, making sure to always follow certain rules:

> - By default, imports are disallowed unless they have been explicitly added to an authorization list by the user

Reviewer: ```suggestion
- By default, imports are disallowed unless they have been explicitly added to an authorization list by the user.
```

>    - Even then because some innocuous packages like `re` can give access to potentially harmful packages as in `re.subprocess`, subpackages that match a list of dangerous patterns are not imported.

Reviewer: ```suggestion
   - Even so, because some innocuous packages like `re` can give access to potentially harmful packages as in `re.subprocess`, subpackages that match a list of dangerous patterns are not imported.
```

 - The total count of elementary operations processed is capped to prevent infinite loops and resource bloating.
 - Any operation that has not been explicitly defined in our custom interpreter will raise an error.

> As a result, this interpreter is safer. We have used it on a diversity of use cases, without ever observing any damage to the environment. 

Reviewer: ```suggestion
As a result, this interpreter is safer. We have used it on a diversity of use cases, without ever observing any damage to the environment.
```

> However this solution is certainly not watertight, since no local python sandbox can really be: one could imagine occasions where LLMs fine-tuned for malignant actions could still hurt your environment.

Reviewer: ```suggestion
However, this solution is certainly not watertight, as no local python sandbox can really be: one could imagine occasions where LLMs fine-tuned for malignant actions could still hurt your environment.
```

> For instance if you've allowed an innocuous package like `Pillow` to process images, the LLM could generate thousands of saves of images to bloat your hard drive.

Reviewer: ```suggestion
For instance, if you have allowed an innocuous package like `Pillow` to process images, the LLM could generate thousands of image saves to bloat your hard drive.
```

Other examples of attacks can be found [here](https://gynvael.coldwind.pl/n/python_sandbox_escape).

Running these targeted malicious code snippet require a supply chain attack, meaning the LLM you use has been intoxicated.

The likelihood of this happening is low when using well-known LLMs from trusted inference providers, but it is still non-zero.

> [!WARNING]
> The only way to run LLM-generated code securely is to isolate the execution from your local environment.

So if you want to exercise caution, you should use a remote execution sandbox.

Here are examples of how to do it.

## Sandbox setup for secure code execution

When working with AI agents that execute code, security is paramount. This guide describes how to set up and use secure sandboxes for your agent applications using either E2B cloud sandboxes or local Docker containers.

### E2B setup

#### Installation

1. Create an E2B account at [e2b.dev](https://e2b.dev)
2. Install the required packages:
```bash

> pip install e2b_code_interpreter smolagents

Reviewer: ```suggestion
pip install 'smolagents[e2b]'
```

            cmd=["python", "-c", code],
            user="nobody"
        )

        # Collect all output
        return exec_result.output.decode() if exec_result.output else None

    def cleanup(self):
        if self.container:
            try:
                self.container.stop()
            except docker.errors.NotFound:
                # Container already removed, this is expected
                pass
            except Exception as e:
                print(f"Error during cleanup: {e}")
            finally:
                self.container = None  # Clear the reference

# Example usage:
sandbox = DockerSandbox()

try:
    # Define your agent code
    agent_code = """
import os
from smolagents import CodeAgent, HfApiModel

# Initialize the agent
agent = CodeAgent(
    model=HfApiModel(token=os.getenv("HF_TOKEN"), provider="together"),
    tools=[]
)

# Run the agent
response = agent.run("What's the 20th Fibonacci number?")
print(response)
"""

    # Run the code in the sandbox
    output = sandbox.run_code(agent_code)
    print(output)

finally:
    sandbox.cleanup()
```

### Best practices for sandboxes

These key practices apply to both E2B and Docker sandboxes:

> #### Resource management
> - Set memory and CPU limits
> - Implement execution timeouts
> - Monitor resource usage
> 
> #### Security
> - Run with minimal privileges
> - Disable unnecessary network access
> - Use environment variables for secrets
> 
> #### Environment
> - Keep dependencies minimal
> - Use fixed package versions
> - If you use base images, update them regularly
> 
> #### Cleanup
> 
> Always ensure proper cleanup of resources, especially for Docker containers, to avoid having dangling containers eating up resources.

Reviewer: Maybe a list of items?
```suggestion
- Resource management
  - Set memory and CPU limits
  - Implement execution timeouts
  - Monitor resource usage
- Security
  - Run with minimal privileges
  - Disable unnecessary network access
  - Use environment variables for secrets
- Environment
  - Keep dependencies minimal
  - Use fixed package versions
  - If you use base images, update them regularly

- Cleanup
  - Always ensure proper cleanup of resources, especially for Docker containers, to avoid having dangling containers eating up resources.
```
